contracts/0.8.9/LidoOracleNew.sol

```diff
@@ -0,0 +1,571 @@
+// SPDX-FileCopyrightText: 2022 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.9;
+
+import { ERC165Checker } from "@openzeppelin/contracts-v4.4/utils/introspection/ERC165Checker.sol";
+import { AccessControlEnumerable } from "@openzeppelin/contracts-v4.4/access/AccessControlEnumerable.sol";
+
+import "./CommitteeQuorum.sol";
+import "./ReportEpochChecker.sol";
+import "./interfaces/ILido.sol";
+import "./interfaces/IBeaconReportReceiver.sol";
+
+
+/**
+ * @title Implementation of an ETH 2.0 -> ETH oracle
+ *
+ * The goal of the oracle is to inform other parts of the system about balances controlled by the
+ * DAO on the ETH 2.0 side. The balances can go up because of reward accumulation and can go down
+ * because of slashing.
+ *
+ * The timeline is divided into consecutive frames. Every oracle member may push its report once
+ * per frame. When the equal reports reach the configurable 'quorum' value, this frame is
+ * considered finalized and the resulting report is pushed to Lido.
+ *
+ * Not all frames may come to a quorum. Oracles may report only to the first epoch of the frame and
+ * only if no quorum is reached for this epoch yet.
+ */
+contract LidoOracleNew is CommitteeQuorum, AccessControlEnumerable, ReportEpochChecker {
+    using ERC165Checker for address;
+    using UnstructuredStorage for bytes32;
+
+    event AllowedBeaconBalanceAnnualRelativeIncreaseSet(uint256 value);
+    event AllowedBeaconBalanceRelativeDecreaseSet(uint256 value);
+    event BeaconReportReceiverSet(address callback);
+
+    event CommitteeMemberReported(
+        uint256 epochId,
+        uint256 beaconBalance,
+        uint256 beaconValidators,
+        address caller,
+        uint256 wcBufferedEther,
+        uint256[] requestIdToFinalizeUpTo,
+        uint256[] finalizationPooledEtherAmount,
+        uint256[] finalizationSharesAmount
+    );
+
+    event ConsensusReached(
+        uint256 epochId,
+        uint256 beaconBalance,
+        uint256 beaconValidators,
+        uint256 wcBufferedEther,
+        uint256[] requestIdToFinalizeUpTo,
+        uint256[] finalizationPooledEtherAmount,
+        uint256[] finalizationSharesAmount
+    );
+
+    event PostTotalShares(
+         uint256 postTotalPooledEther,
+         uint256 preTotalPooledEther,
+         uint256 timeElapsed,
+         uint256 totalShares
+    );
+
+    event ContractVersionSet(uint256 version);
+
+
+    struct MemberReport {
```

Comment:
could be packed more tightly:
`uint256 beaconValidators` → `uint96 beaconValidators` (validator index is uint64)
`uint256 epochId` → `uint64 epochId` (one epoch lasts ~6.4 minutes)
`uint256[] nodeOperatorsWithExitedValidators` → `uint64[] ...`
`uint256 wcBufferedEther` → `uint96 wcBufferedEther`
`uint256 newDepositBufferWithdrawalsReserve` → `uint96 newDepositBufferWithdrawalsReserve`